contracts/trading/seaport/zones/immutable-signed-zone/v2/ImmutableSignedZoneV2.sol

```diff
@@ -0,0 +1,584 @@
+        address recoveredSigner = ECDSA.recover(digest, bytes32(signature[0:32]), bytes32(signature[32:64]));
+
+        // Revert if the signer is not active.
+        // This also reverts if the digest constructed on serverside is incorrect.
+        if (!_signers[recoveredSigner].active) {
+            revert SignerNotActive(recoveredSigner);
+        }
+
+        // All validation completes and passes with no reverts, return valid.
+        validOrderMagicValue = ZoneInterface.validateOrder.selector;
+    }
+
+    /**
+     * @dev Get the supported substandards of the contract.
+     *
+     * @return substandards Array of substandards supported.
+     */
+    function _getSupportedSubstandards() internal pure returns (uint256[] memory substandards) {
+        // support substandards 3, 4 and 6
+        substandards = new uint256[](3);
+        substandards[0] = 3;
+        substandards[1] = 4;
+        substandards[2] = 6;
+    }
+
+    /**
+     * @dev Derive the signedOrder hash from the orderHash and expiration.
+     *
+     * @param fulfiller        The expected fulfiller address.
+     * @param expiration       The signature expiration timestamp.
+     * @param orderHash        The order hash.
+     * @param context          The optional variable-length context.
+     * @return signedOrderHash The signedOrder hash.
+     */
+    function _deriveSignedOrderHash(address fulfiller, uint64 expiration, bytes32 orderHash, bytes calldata context)
+        internal
+        view
+        returns (bytes32 signedOrderHash)
+    {
+        // Derive the signed order hash.
+        signedOrderHash =
+            keccak256(abi.encode(_SIGNED_ORDER_TYPEHASH, fulfiller, expiration, orderHash, keccak256(context)));
+    }
+
+    /**
+     * @dev Validate substandards 3, 4 and 6 based on context.
+     *
+     * @param context        Bytes payload of context.
+     * @param zoneParameters The zone parameters.
+     */
+    function _validateSubstandards(bytes calldata context, ZoneParameters calldata zoneParameters) internal pure {
+        uint256 startIndex = 0;
+        uint256 contextLength = context.length;
+
+        // Each _validateSubstandard* function returns the length of the substandard
+        // segment (0 if the substandard was not matched).
+
+        if (startIndex == contextLength) return;
+        startIndex = _validateSubstandard3(context[startIndex:], zoneParameters) + startIndex;
+
+        if (startIndex == contextLength) return;
+        startIndex = _validateSubstandard4(context[startIndex:], zoneParameters) + startIndex;
+
+        if (startIndex == contextLength) return;
+        startIndex = _validateSubstandard6(context[startIndex:], zoneParameters) + startIndex;
+
+        if (startIndex != contextLength) {
+            revert InvalidExtraData("invalid context, unexpected context length", zoneParameters.orderHash);
+        }
+    }
+
+    /**
+     * @dev Validates substandard 3.
```

Comment:
Do you think it might be worth to add brief 1-liners about these substandards to give the reader more context? Or do you think its reasonable to assume the reader will have the SIP docs up? The latter could well be true

Reply:
https://github.com/immutable/contracts/pull/189/commits/eb5495af213b142c9edbca0605da13ed221ab8c7